### PR TITLE
fix issue 6549 - Implement contracts without implementation

### PR DESCRIPTION
**Pull request by ibuclaw**
Supersedes #3611

### CODE REVIEW

File: test/fail_compilation/fail12901.d [hunk 1]
```diff
@@ -1,14 +0,0 @@
...
----
-*/
-
-struct S
-{
-    int a;
-    this(int n)
-    in { a = n; }
-    // no body
-}
```
Comment (wilzbach) [inline test/fail_compilation/fail12901.d]: What's the motivation for removing this? Shouldn't it still be forbidden?

File: test/runnable/testcontracts.d [hunk 2]
```diff
@@ -421,6 +421,56 @@ void test6417()
...
+    abstract int foo(int)
+    in { ++icount; }
+    out { ++ocount; }
+}
+
+class CD6549 : C6549
+{
+    override int foo(int)
+    in { assert(false); }
+    body { return 10; }
```
Comment (BBasile) [inline test/runnable/testcontracts.d]: use `do` since soon or later someone will have to patch the test.

File: test/runnable/testcontracts.d [hunk 3]
```diff
@@ -421,6 +421,56 @@ void test6417()
...
+    int foo(int)
+    in { ++icount; }
+    out { ++ocount; }
+}
+
+class DD6549 : D6549
+{
+    override int foo(int)
+    in { assert(false); }
+    body { return 10; }
```
Comment (BBasile) [inline test/runnable/testcontracts.d]: use `do`.

File: test/fail_compilation/fail12901.d (hunk 1 above)
Comment (ibuclaw) [reply test/fail_compilation/fail12901.d]: Not sure as this is would be the third person to rebase and submit this patch.

It certainly seems reasonable to me to not include body-less ctors, dtors, and postblits to have contracts in the relaxation of this rule.

Comment (wilzbach) [reply test/fail_compilation/fail12901.d]: Hmm according to the comment:

> Contracts can only appear without a body when they are virtual interface functions or abstract

and its respective if condition, this should still be forbidden.
I just checked locally and it doesn't throw an error anymore.
The problem is this jump:

```d
if (StructDeclaration sd = parent.isStructDeclaration())
{
    if (funcdecl.isCtorDeclaration())
    {
        goto Ldone;
     }
}
```

Comment (ibuclaw) [reply test/fail_compilation/fail12901.d]: Ah, interesting. I wonder the reason for that.

**Comment by andralex** [comment]
@ibuclaw no mercy

**Comment by wilzbach** [comment]
Ehm you just broke [12901](https://issues.dlang.org/show_bug.cgi?id=12901) without an argument for doing so. The issue is titled like the test case that was removed:

> `in`/`out` contracts on struct constructor must require function body

**Comment by ibuclaw** [comment]
Right, I don't remember seeing anyone actually approve this, and was waiting until I had sorted that test out.

**Comment by quickfur** [comment]
Huh, issue 12901 is really more of a problem of allowing arbitrary assignments inside the contract body.  The original fix was invalid, since it doesn't stop you from writing `this(int n) in { a = n; } do {}`, which is still contract abuse. I'm not sure what the correct fix should be, perhaps require the contracts be pure or something.

**Pull request merged**